src/protocol/sort/secureapplyinv.rs

```diff
@@ -42,35 +42,45 @@ impl SecureApplyInv {
     /// 5. All helpers call `apply` to apply the permutation locally.
     pub async fn execute<F: Field>(
         ctx: ProtocolContext<'_, Replicated<F>, F>,
-        input: Vec<Replicated<F>>,
-        sort_permutation: Vec<Replicated<F>>,
+        input: &mut [Replicated<F>],
+        sort_permutation: &mut [Replicated<F>],
     ) -> Result<Vec<Replicated<F>>, BoxError> {
-        let random_permutations = get_two_of_three_random_permutations(input.len(), &ctx.prss());
+        let (left_random_permutation, right_random_permutation) =
+            get_two_of_three_random_permutations(input.len(), &ctx.prss());
 
         let (mut shuffled_input, shuffled_sort_permutation) = try_join(
-            Shuffle::new(input, random_permutations.clone()).execute(ctx.narrow(&ShuffleInputs)),
-            Shuffle::new(sort_permutation, random_permutations)
-                .execute(ctx.narrow(&ShufflePermutation)),
+            shuffle_shares(
+                input,
+                &left_random_permutation,
+                &right_random_permutation,
+                ctx.narrow(&ShuffleInputs),
+            ),
+            shuffle_shares(
+                sort_permutation,
+                &left_random_permutation,
+                &right_random_permutation,
+                ctx.narrow(&ShufflePermutation),
+            ),
         )
         .await?;
         let revealed_permutation =
             reveal_permutation(ctx.narrow(&RevealPermutation), &shuffled_sort_permutation).await?;
         // The paper expects us to apply an inverse on the inverted Permutation (i.e. apply_inv(permutation.inverse(), input))
         // Since this is same as apply(permutation, input), we are doing that instead to save on compute.
-        apply(revealed_permutation, &mut shuffled_input);
+        apply_inv(&revealed_permutation, &mut shuffled_input);
         Ok(shuffled_input)
```

Comment:
Since we are taking a mutable ref of input, we will have more memory taken at the given time (input, sort_permutation, shuffled_input and shuffled_sort_permutation are all in memory when this function is executed). Even after this function has returned, input and shuffled_input are taking memory (since input is not consumed by the caller). Dont think this is ideal

Reply:
I see what you're saying. Previously, the memory for these things was de-allocated sooner.

`input` and `shuffled_sort_permutation` were previously consumed on line 51, and need not occupy memory from that point onwards. In this version that's no longer the case. They remain in memory until the end of the calling function, when `input` is no longer referenced. 

That's a good point. I guess I didn't really need to mess with these did I. I really just wanted to eliminate the `clone` on line 51. Let me see if I can bring them back.